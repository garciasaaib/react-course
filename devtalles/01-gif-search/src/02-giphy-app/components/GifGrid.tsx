import { GifItem } from "./";
import { useFetchGifs } from "../hooks/useFetchGifs";

export const GifGrid = ({ category }: { category: string }) => {
	const { gifs, loading } = useFetchGifs(category);

	return (
		<>
			<h3>{category}</h3>
			{loading && <h2>Cargando...</h2>}

			<div className="card-grid">
				{gifs.map((image) => (
					<GifItem key={image.id} {...image} />
				))}
			</div>
		</>
	);
};
