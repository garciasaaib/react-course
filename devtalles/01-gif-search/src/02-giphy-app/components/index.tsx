export { AddCategory } from './AddCategory';
export { GifGrid } from './GifGrid';
export { GifItem } from './GifItem';