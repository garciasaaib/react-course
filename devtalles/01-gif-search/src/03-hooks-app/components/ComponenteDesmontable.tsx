import { useEffect, useState } from "react"

export const ComponenteDesmontable = () => {
  const [coords, setCoords] = useState({x:0, y:0})
  // reglas de montaje y desmontaje
  useEffect(() => {
    console.log('Message Mounted')
    return () => {
      console.log('Message Unmounted')
    }
  }, [])

  // efecto encadenado por el movimiento del mouse
  useEffect(() => {
    // creamos la funcion
    const onMouseMove = ({x, y}: MouseEvent) => {
      const coords = {x,y}
      setCoords(coords)
    }

    // agregamos esta funcion por referencia cuando se monta
    window.addEventListener('mousemove', onMouseMove)

    // quitamos la funcion del evento por referencia
    return () => {
      window.removeEventListener('mousemove',onMouseMove)
    }
  },[])
  return (
    <>
      <h3>Usuario ya existe</h3>
      <p>x: {coords.x}</p>
      <p>y: {coords.y}</p>
    </>
  )
}
