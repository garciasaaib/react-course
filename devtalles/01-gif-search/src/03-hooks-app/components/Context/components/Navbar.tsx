import { Link, useLocation } from "react-router";

export const Navbar = () => {
  const {pathname} = useLocation()
  const onIsActive = (thisPath: string) => {
    return pathname === thisPath ? " underline" : "";

  }
	return (
		<nav className="flex gap-1">
			<Link className={`hover:text-white${onIsActive('/')}`} to="/">
				Home
			</Link>
			<Link className={`hover:text-white${onIsActive('/about')}`} to="/about">
				About
			</Link>
			<Link className={`hover:text-white${onIsActive('/login')}`} to="/login">
				Login
			</Link>
		</nav>
	);
};
