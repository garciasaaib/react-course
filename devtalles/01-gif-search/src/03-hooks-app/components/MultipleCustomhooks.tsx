export const MultipleCustomhooks = () => {
  return (
    <div>MultipleCustomhooks</div>
  )
}