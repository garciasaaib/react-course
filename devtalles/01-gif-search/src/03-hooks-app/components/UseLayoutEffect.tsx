
// useLayoutEffect es un useEffect que se dispara despues de que el componente es montado
// y despues de los useEffect

// import { useLayoutEffect } from "react"
import { useLayoutEffect, useRef, useState } from "react";
import { useCounter } from "../hooks/useCounter";
import { useFetch } from "../hooks/useFetch";
import { BreakingBadAPIResponse } from "../types/Breakingbadapi";

// es usado para enfocar los cambios en algun componente
export const UseLayoutEffect = () => {
	const { counter, decrement, increment, reset } = useCounter(1);
	const { isLoading, data } = useFetch<BreakingBadAPIResponse[]>(
		`https://api.breakingbadquotes.xyz/v1/quotes/${counter}`
	);
	return (
		<div>
			<h1>BreakingBad Quotes</h1>
			{isLoading && <div className="alert alert-info text-center">
        Loading...
    </div>
		}

			<button disabled={counter <= 1} onClick={decrement} className="btn">Anterior</button>
      <button onClick={reset} className="btn">Reset</button>
			<button onClick={increment} className="btn">Siguiente</button>
			{data && <Quote author={data[0].author} quote={data[0].quote} /> }
		</div>
	);
}



export const Quote = ({ author, quote }: {author:string, quote:string}) => {

	// hacemos referencia al tag p para que no se renderize todo despues de su cambio en quote
  const pRef = useRef<HTMLParagraphElement>(null);
  const [boxSize, setBoxSize] = useState({ width: 0, height: 0 })

	// este efecto se disparara cada vez que quote cambie
  useLayoutEffect(() => {
		// useLayoutEffect nos ayuda a identificar que la referencia ya esta asignada, pero por TS no lo permite.
    if (pRef.current) {
			const { height, width } =  pRef.current.getBoundingClientRect();
			setBoxSize({ height, width });
		}
  }, [quote])
  


  return (
    <>
      <blockquote 
        className="blockquote text-end flex flex-col"
      >
          <p ref={ pRef } className="mb-1">{ quote }</p>
					<br/>
          <footer className="blockquote-footer"> { author } </footer>
      </blockquote>

      <code>{ JSON.stringify(boxSize) }</code>
    </>
  )
}