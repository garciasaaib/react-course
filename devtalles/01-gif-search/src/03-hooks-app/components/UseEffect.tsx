import { ChangeEvent, useEffect, useState } from "react";
import { ComponenteDesmontable } from "./ComponenteDesmontable";

export const UseEffect = () => {
  const [state, setState] = useState({
    username: 'Adrian',
    email: 'adrian@gmail.com',
  })
  const {username, email} = state
  const onInputChange = ({target}: ChangeEvent<HTMLInputElement>) => {
    const {value, name} = target
    setState({...state, [name]: value})
  }
  useEffect(() => {
    // body: all that this does


    // return () => {
      // clean up: cleanning for unmount component
    //}
  }, []) // dependencies: watching variables
	return (
		<>
			<h1>SimpleForm</h1>
			<hr />
			<form>
				<input onChange={onInputChange} value={username} type="text" placeholder="Username" name="username" />
				<input onChange={onInputChange} value={email} type="email" placeholder="Email" name="email" />
			</form>
      {/* El monte y desmonte se genera desde el compoonente padre */}
      { username === 'Adrian' && <ComponenteDesmontable /> }
		</>
	);
};
