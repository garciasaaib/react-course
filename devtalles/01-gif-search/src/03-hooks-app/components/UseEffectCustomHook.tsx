import { ComponenteDesmontable } from "./ComponenteDesmontable";
import { useEffectCustomHook } from "../hooks/useEffectCustomHook";

const INITIAL_VALUES = {
	username: "",
	email: "",
	password: "",
};
export const UseEffectCustomHook = () => {
	const { state, onInputChange, onInputReset } = useEffectCustomHook({ ...INITIAL_VALUES });

	return (
		<>
			<h1>SimpleForm</h1>
			<hr />
			<form>
				<input
					onChange={onInputChange}
					value={state.username}
					type="text"
					placeholder="Username"
					name="username"
				/>
				<input
					onChange={onInputChange}
					value={state.email}
					type="email"
					placeholder="Email"
					name="email"
				/>
				<input
					onChange={onInputChange}
					value={state.password}
					type="password"
					placeholder="Password"
					name="password"
				/>
			</form>
      <button className="btn" onClick={onInputReset}>Reset</button>
			{/* El monte y desmonte se genera desde el compoonente padre */}
			{state.username === "Adrian" && <ComponenteDesmontable />}
		</>
	);
};
