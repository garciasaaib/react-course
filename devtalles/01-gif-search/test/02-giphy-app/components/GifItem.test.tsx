import React from "react";
import { render, screen } from "@testing-library/react";
import { GifItem } from "../../../src/02-giphy-app/components/GifItem";

describe("<GifItem />", () => {
	test("Debe crear en snapshoot de GifItem", () => {
		const props = {
			title: "Saitama",
			url: "https://one-punch.com/saitama.jpg",
			id: "1",
		};
		const { container } = render(<GifItem {...props} />);
		expect(container).toMatchSnapshot();
	});

	test("should show the image with url and alt", () => {
		const props = {
			title: "Saitama",
			url: "https://one-punch.com/saitama.jpg",
			id: "1",
		};
		render(<GifItem {...props} />);
		// screen.debug()
    const {src, alt} = screen.getByRole<HTMLImageElement>('img')
    expect(src).toBe(props.url)
    expect(alt).toBe(props.title)
	});

  test("should show the title", () => {
		const props = {
			title: "Saitama",
			url: "https://one-punch.com/saitama.jpg",
			id: "1",
		};
		render(<GifItem {...props} />);
		// screen.debug()
    // const {src, alt} = screen.getByRole<HTMLImageElement>('img')
    // expect(src).toBe(props.url)
    expect(screen.getByText(props.title)).toBeTruthy()
	});
});
