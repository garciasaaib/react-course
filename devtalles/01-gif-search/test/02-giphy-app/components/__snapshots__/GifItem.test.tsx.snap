// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GifItem /> Debe crear en snapshoot de GifItem 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="Saitama"
      src="https://one-punch.com/saitama.jpg"
    />
    <p>
      Saitama
    </p>
  </div>
</div>
`;
